src/libcore/task/wake.rs

```diff
@@ -63,21 +63,99 @@ pub struct RawWakerVTable {
     /// required for this additional instance of a [`RawWaker`] and associated
     /// task. Calling `wake` on the resulting [`RawWaker`] should result in a wakeup
     /// of the same task that would have been awoken by the original [`RawWaker`].
-    pub clone: unsafe fn(*const ()) -> RawWaker,
+    clone: unsafe fn(*const ()) -> RawWaker,
 
     /// This function will be called when `wake` is called on the [`Waker`].
     /// It must wake up the task associated with this [`RawWaker`].
     ///
     /// The implemention of this function must not consume the provided data
     /// pointer.
-    pub wake: unsafe fn(*const ()),
+    wake: unsafe fn(*const ()),
 
     /// This function gets called when a [`RawWaker`] gets dropped.
     ///
     /// The implementation of this function must make sure to release any
     /// resources that are associated with this instance of a [`RawWaker`] and
     /// associated task.
-    pub drop: unsafe fn(*const ()),
+    drop: unsafe fn(*const ()),
+}
+
+impl RawWakerVTable {
+    /// Creates a new `RawWakerVTable` from the provided `clone`, `wake`, and
+    /// `drop` functions.
+    ///
+    /// # `clone`
+    ///
+    /// This function will be called when the [`RawWaker`] gets cloned, e.g. when
+    /// the [`Waker`] in which the [`RawWaker`] is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a [`RawWaker`] and associated
+    /// task. Calling `wake` on the resulting [`RawWaker`] should result in a wakeup
+    /// of the same task that would have been awoken by the original [`RawWaker`].
+    ///
+    /// # `wake`
+    ///
+    /// This function will be called when `wake` is called on the [`Waker`].
+    /// It must wake up the task associated with this [`RawWaker`].
+    ///
+    /// The implemention of this function must not consume the provided data
+    /// pointer.
+    ///
+    /// # `drop`
+    ///
+    /// This function gets called when a [`RawWaker`] gets dropped.
+    ///
+    /// The implementation of this function must make sure to release any
+    /// resources that are associated with this instance of a [`RawWaker`] and
+    /// associated task.
+    pub const fn new(
```

Comment:
cc @oli-obk, this ain't gonna work once we make this module... `#[stable(...)]`, confirm?

Reply:
Confirmed:
```rust
#![feature(staged_api)]

#[stable(feature = "foo", since = "1.0.0")]
const fn foo(drop: unsafe fn(*const ())) {}
fn main() {}
```
==>
```
error[E0723]: function pointers in const fn are unstable (see issue #57563)
```

Reply:
I think an alternative would be to use `#[non_exhaustive]` and `Default`, it would lose the “must specify at least  these 3 functions” of the constructor, but this is such a rarely used API that that could be ok.

Reply:
I also think this constructor is not that useful. The type can already be constructed without it (all fields are public).

Reply:
@oli-obk this PR changes all the struct fields to private.

Reply:
Surely we can make the module stable and include unstable uses of `const fn` in it?

Reply:
@cramertj Yup, it compiles fine, confirmed with:
```rust
#![feature(staged_api)]

#![stable(feature = "foo", since = "1.0.0")]

fn main() {}

#[stable(feature = "foo", since = "1.0.0")]
mod foo {
    #[unstable(feature = "bar", issue = "0")]
    const fn foo(drop: unsafe fn(*const ())) {}
}
```

What you cannot do is use an unstable `const fn` inside a stable one; all stable `const fn`s must adhere to the `min_const_fn` regime we set up and only call other `min_const_fn`s.

Reply:
Oh-- that doesn't work here-- we need to be able to call this from const contexts on stable. What's the blocker in the implementation here? This seems like a reasonable thing to add to me.

Reply:
(by "unstable uses of `const fn`" i meant that the function would be stable but would use unstably-const features, which I realize now was ambiguous)

Reply:
That's not possible due to the way we set up `const fn` checks. Stable const fn can only do stable const fn things and call other stable const fn. There is no hiding of instability, because right now we are basically stabilizing the *body* of a `const fn` together with its API.

After the `impl const Trait` rfc is merged we can make an RFC for the "future work" section of function pointers (https://github.com/rust-lang/rfcs/pull/2632/files#diff-93d98aeffe5c71a78e51f4c4dbdd1a4bR549). Or I can directly include it in the RFC since it's pretty much a trivial addition (the only question is a small syntax bikeshed).

We can't stabilize a function like your proposed `new` method, because it's not clear whether the API you've given it exposes that you are going to call the function pointer inside, or that Rust guarantees that you will not call it.

Reply:
Me, @cramertj, and @oli-obk [discussed this on Discord](https://discordapp.com/channels/442252698964721669/443151225160990732/555099673245777920). Our conclusion was that we can get away with special casing the two functions with an attribute in the style of `#[rustc_promotable]` as used in `librustc_mir/transform/qualify_min_const_fn.rs` under some unwieldy name `#[rustc_const_allow_fn_pointer_only_for_raw_wakers]`.

This new attribute would only be used for those functions and would give users on stable the ability to *call* the functions and pass function pointers to them such that the wakers can be stored inside `static` items. This would commit us to *a* solution for passing function pointers which I think everyone is agreed we want to allow at some point. Therefore it is deemed to be of low risk. However, we will not stabilize the ability to *define* `const fn`s which take `fn` pointers at this time since that would lock us into a specific model and semantics for the surface syntax.

Introducing this attribute should be done in a separate PR and should then be FCP-accepted by T-Lang. I *believe* that this PR can proceed independently of that once since the attribute is only necessary for stability purposes.

Reply:
Note that I've now also tagged these functions as `#[rustc_promotable]` which is necessary in order to use them in many of the existing sites that were previously creating instances of the structs (which may e.g. capture generics which isn't possible in named statics). I believe this is a sound usage of the attribute, but I'd like @oli-obk and @RalfJung to confirm (the function just takes a set of function pointers and puts them in a struct).

Reply:
The long-term plan is to not ever promote calls to `const fn` implicitly, and instead require users to do that explicitly.  I had hoped we wouldn't mark any new things as promotable, given that we realized that promoting `const fn` is a huge can of worms and probably was a mistake to begin with.

So, I'm not a fan. But if y'all are aware of the context and deem this important enough to be an exception, I won't object.

Reply:
Such constructor functions seem fine. I'm unhappy with extending the set of promotable funcitons out of the reasons stated by @RalfJung. That said, I think we can actually have a sane way forward for the attribute, but I'll discuss this in https://github.com/rust-rfcs/const-eval/issues/19 in order to not derail this PR.

So: I'm okay with doing this, but be aware that this increases technical debt.

Reply:
This is another reason for my position that "cheating" with "exposed bodies" is actually good overall, as it significantly reduces complexity while allowing maximal features but still keeping "significance of `fn` bodies to stability" opt-in.

Reply:
@cramertj would it be too horrible for users if the functions would not have `#[rustc_promotable]`, but instead are required to be used via an intermediate constant?

```rust
const fn foo() -> i32 { 42 }
fn main() {
    const HELPER: i32 = foo();
    let x: &'static i32 = &HELPER;
}
```

is legal on stable Rust.

Reply:
Provided it works, an intermediate constant seems preferable to an increase of technical debt. The number of people who are going to be using the raw waker API are not that large so a paper cut here seems fine. A user-space macro can also paper over it.

Reply:
If we used an actual trait instead of a struct, it'd be much easier to add new default methods in the future. (I still don't really see how a struct vtable is better than a trait...)

Reply:
> which may e.g. capture generics which isn't possible in named statics

Ah nevermind, I read this earlier but forgot about it again.

You can emulate generic constants with other const fn:

```rust
#![feature(const_fn)]

fn main() {
}

struct Foo<T> {
    f: fn(T),
}

fn foo<T>(_: T) {}

const fn const_generic<T: 'static>() -> &'static Foo<T> {
    &Foo {
        f: foo,
    }
}

fn generic<T: 'static>() {
    let x: &'static Foo<T> = const_generic();
}
```

Reply:
> If we used an actual trait instead of a struct, it'd be much easier to add new default methods in the future. (I still don't really see how a struct vtable is better than a trait...)

Has this come up in the RFC before?

Reply:
It's in the RFC, but the reasons aren't clear (to me, at least). https://github.com/rust-lang/rfcs/blob/master/text/2592-futures.md#waking-up

> This mechanism is chosen in favor of trait objects since it allows for more flexible memory management schemes. RawWaker can be implemented purely in terms of global functions and state, on top of reference counted objects, or in other ways. This strategy also makes it easier to provide different vtable functions that will perform different behaviors despite referencing the same underlying wakeable object type.

I bring it up since the reasons don't seem to clearly favor a struct, and we're having issues with making this constructor, that using a trait wouldn't have.

Reply:
Curious. I see all of these as solveable via trait objects, too. Worst case you're going to end up with an additional indirect function pointer call inside the trait impl.

The reduced complexity of using real trait objects feels like a huge advantage to me.

Reply:
The "What does the [xxx] phase require?" sections in https://boats.gitlab.io/blog/post/wakers-i/ seems to have more details. It looks like one sticking point is that the waker must be `Clone`-able. But there's more there and I haven't fully worked through it.

Reply:
Using a vtable rather than a trait makes it easier to swap out one or more of the functions without having some perverse transmutes into marker types.

Reply:
(for that matter, we should be adding `const fn` accessors as well, but that can go in another change)

Reply:
Just in case, I threw together a commit that uses a trait instead of the vtable: https://github.com/rust-lang/rust/compare/master...seanmonstar:raw-wake-as-trait

Reply:
> You can emulate generic constants with other const fn:

That sounds much nicer than adding a new promotable function.

Reply:
I mean, I don't agree that it's nicer-- it seems like a non-obvious dirty hack to me. But I can certainly make that change.

Reply:
That seems rather good? I don't see why perverse transmutes would arise from this?

You can recover the explicit dictionary encoding of the trait by simply having the struct yourself and then delegating to those member function pointers... https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d452b451d098ee65cb90b033dbb33e79

Reply:
Please, I'd like to strongly request that we not take this as an opportunity to reconsider these decisions that were already made on the RFC thread after *extensive* discussion. Turning this into a trait is not an option.

Reply:
I've already explained in full detail in *many* places why this shouldn't be a trait, and I'm not interested in doing it again here.

Reply:
@cramertj I'm sorry! I couldn't find the details, would it help if they were in RFC?

Reply:
The RFC contains this relevant text:

> This mechanism is chosen in favor of trait objects since it allows for more flexible memory management schemes. RawWaker can be implemented purely in terms of global functions and state, on top of reference counted objects, or in other ways. This strategy also makes it easier to provide different vtable functions that will perform different behaviors despite referencing the same underlying wakeable object type.

[here.](https://github.com/rust-lang/rfcs/blob/master/text/2592-futures.md#waking-up)

Reply:
Linking to the discussion:

https://github.com/rust-lang/rfcs/pull/2592#discussion_r233669955

> 

LocalWaker and Waker are themselves trait objects. Both work out to something like *const UnsafeWake. Implementing the Wake trait gives you an implementation of UnsafeWake for Arc<YourType>.

> This API can be awkward to work with in some edge cases though where you want to change the vtable without changing the underlying pointed to type. @aturon is working on a modification of the proposal, and I've suggested considering that we use a more C-like manual vtable instead. This would allow us more flexibility, make type-punning-to-change-the-vtable unnecessary, and removes the hairy questions around whether or not *const UnsafeWake must have a valid vtable.

https://github.com/rust-lang/rfcs/pull/2592#issuecomment-438898492

>  In particular, it seems a bit clearer how to get a Waker from a Wake instance. (Note: the reason we can't provide a direct constructor is that Waker needs to live in core while Wake lives in std).

> Making the vtable explicit makes a lot of sense; it's basically what UnsafeWake is trying to do anyway, but expressed more clearly and sidestepping object safety issues. Edit: and I forgot to mention, that would make it feasible to stabilize the full API surface immediately.

The main discussion about this (that we're reopening here) is by @carllerche  in https://github.com/rust-lang/rfcs/pull/2592#discussion_r252443404

The discussion was completely ignored without providing even an example (except a hypothetical one where you want to hot swap methods without changing the underlying type) that a manual vtable scheme (and all its problems) is actually necessary.

Thus I feel that it's perfectly alright to reopen the discussion about this detail here. Especially in the light of implementation problems of such a manual vtable.

Reply:
Note that I'm absolutely sympathetic to the argument that this is not something that users will really see and implementing it safely isn't possible anyway, so we can just do this all manually. This argument would imo also call for not doing any convenience hacks like allowing promotion, since there are workarounds (generic const function returning a promoted value).

Basically all I'm asking for is an example to be added to the rationale where the vtable is actually modified.

For the record, here's a "real world" waker implementation in one of our projects:

https://github.com/embed-rs/stm32f7-discovery/blob/e2bf713263791c028c2a897f2eb1830d7f09eceb/src/task_runtime.rs#L97-L125

I basically ended up writing all the hypothetical trait methods as inherent methods and then transmuted them to match the `RawWakerVtable` API. If the intent was to reduce transmutes or other unsafe hacks, for me personally this was not achieved. It does feel like an optimization for the rare case instead of the common case. Although I realize I might just be using the API very wrongly.

Reply:
@oli-obk as I read the conversation, the conclusion from the discussion is that you **can** implement the exact same strategies with a trait object compared to the proposed raw vtable, however it may require unsafe / transmute hacks. 

The RFC text still makes the claim that:

> This mechanism is chosen in favor of trait objects since it allows for more flexible memory management schemes.

But I (still) do not believe it is true (no provided example).

The rest of the text:

> This strategy also makes it easier to provide different vtable functions that will perform different behaviors despite referencing the same underlying wakeable object type.

This may be true in theory, but I don't believe it is a huge win in **practice**. Both strategies require significant unsafe code.

Reply:
So... I've been convinced that we should go with the handrolled vtable scheme. There are `Waker`s which are actually borrowing some data in "that specific fat pointer instance". So e.g. the `drop` function needs to be a noop, while the concrete type is still the same in order for the `clone` function to produce a clone of the concrete type and not the reference. One example is to be found [here](https://github.com/rust-lang-nursery/futures-rs/blob/79d667a7187311518c408f4b360f5987568f7c1c/futures-util/src/task/waker_ref.rs#L58). I've been told by @cramertj  there are several more in fuchsia. You can find the entire discussion on discord [here](https://discordapp.com/channels/442252698964721669/443151225160990732/555776382932746260)

While I am still not happy that we're extending the set of `rustc_promotable` functions, @cramertj says that requiring users to write

```rust
const fn helper<T>() -> &'static RawWakerVtable {
    &RawWakerVtable::new(stuff here)
}
let vtable = helper();
```

instead of

```rust
let vtable = &RawWakerVtable::new(stuff here);
```

is not really discoverable and messy to explain. I'm trusting their judgement on this and and also think that this one pure constructor function is ok to make promotable since it literally does no operations, it just constructs an aggregate. Some theoretical thoughts about why this one function could be stably defined as promotable in the future can be found in https://github.com/rust-rfcs/const-eval/issues/19#issuecomment-472799062

Reply:
> While I am still not happy that we're extending the set of rustc_promotable functions, @cramertj says that requiring users to write

Note that users could have it a bit simpler, they could write
```rust
const VTABLE: RawWakerVtable = RawWakerVtable::new(stuff here);
let vtable = &VTABLE;
```

That's much less messy. If you just replace all uses of `vtable` by `&VTABLE`, you didn't even add a single line of code.

But I also agree that this function is certainly always fine to promote. My counter-argument is that making this promotable now means we'll have more code that needs to be migrated to the new form, when/if we are eventually able to write
```rust
let vtable = const { &RawWakerVtable::new(stuff here) };
```
Of course, the is assuming `const` blocks is the way we go. I'll leave it up to the lang team to weigh "slightly easier construction of static vtables now" against "slightly less code that has to be migrated away from a deprecated pattern later".

Reply:
> > 
> 
> Note that users could have it a bit simpler, they could write
> 
> ```rust
> const VTABLE: RawWakerVtable = RawWakerVtable::new(stuff here);
> let vtable = &VTABLE;
> ```

That seems entirely fine to me given that it can be easily papered over with a macro and that few people will need to write this a lot.

Reply:
> That's much less messy. If you just replace all uses of vtable by &VTABLE, you didn't even add a single line of code.

That doesn't work if you are initializing the vtable by depending on generic parameters:

```rust
fn mk_some_waker_thingy<T>() -> WakerThingy {
    // the next line doesn't work, because it depends on generic parameters
    const VTABLE: RawWakerVtable = RawWakerVtable::new(some_fn::<T>, ...);
    // so we need to work around this issue via generic const fns:
    const fn mk_vtable<T>() -> &'static RawWakerVtable {
        &RawWakerVtable::new(some_fn::<T>, ...)
    }
    let vtable = mk_vtable();
}
```

Reply:
Ah, dang.

The only other proposal I can come up with is to have a `const fn` in libstd that returns an `&'static RawWakerVtable`. Then not every user has to write such a helper themselves. That should work, right?

But I can also live with making this promoteable, as mentioned before.

Reply:
That wouldn't work either, because then you could potentially call that const fn at runtime... and we can't promote runtime arguments.

Reply:
All right, I give up. Sorry for wasting your time.^^

Reply:
You cannot do that in generic contexts. You cannot declare an explicit `const` or `static` because those cannot rely on generic parameters.

Reply:
You can make a `const` that depends on generic parameters via associated `const` in an inherent `impl`, FWIW.

(I just want `const FOO<...>` too tho)